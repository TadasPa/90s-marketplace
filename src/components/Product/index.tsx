import React, { FC, useCallback } from "react";
import IProduct from "../../types/product";

interface IProductDetails {
  product: IProduct;
  onAdd: (product: IProduct) => void;
}

const Product: FC<IProductDetails> = ({ product, onAdd }) => {
  const { title, price, img } = product;
  const handleAdd = useCallback(() => onAdd(product), [product, onAdd]);

  return (
    <div>
      <h1>{title}</h1>
      <p>Price: {price}</p>

      <button
        data-testid={`product-add-to-cart-${product.id}-testid`}
        onClick={handleAdd}
      >
        Add to cart
      </button>

      <div>
        <img src={img} alt={title} width={640} />
      </div>
    </div>
  );
};

export default Product;
