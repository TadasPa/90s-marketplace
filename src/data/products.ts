import pictureA from '../assets/images/a.jpg';
import pictureB from '../assets/images/b.jpg';
import IProduct from '../types/product';

export const PRODUCT_A: IProduct = { id: 'a', title: 'Product A', price: '10USD', img: pictureA };
export const PRODUCT_B: IProduct = { id: 'b', title: 'Product B', price: '30USD', img: pictureB };
