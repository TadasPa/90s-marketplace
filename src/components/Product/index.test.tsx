import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Product from ".";
import { PRODUCT_A, PRODUCT_B } from "../../data/products";

test("renders Product successfully", () => {
  const { container } = render(
    <Product product={PRODUCT_A} onAdd={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});

test("calls add callback", () => {
  const handleAddMock = jest.fn();
  const addButtonTestId = `product-add-to-cart-${PRODUCT_B.id}-testid`;

  render(<Product product={PRODUCT_B} onAdd={handleAddMock} />);

  userEvent.click(screen.getByTestId(addButtonTestId));
  expect(handleAddMock).toHaveBeenCalledWith(PRODUCT_B);
});
