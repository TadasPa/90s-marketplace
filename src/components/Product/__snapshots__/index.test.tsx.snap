// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Product successfully 1`] = `
<div>
  <div>
    <h1>
      Product A
    </h1>
    <p>
      Price: 
      10USD
    </p>
    <button
      data-testid="product-add-to-cart-a-testid"
    >
      Add to cart
    </button>
    <div>
      <img
        alt="Product A"
        src="a.jpg"
        width="640"
      />
    </div>
  </div>
</div>
`;
